### PR TITLE
Fix: `no-useless-concat` only check useless concat of literals on actual string concat (fixes #3575)

### LINKED ISSUE #3575: Rule "no-useless-concat" reports false positive
ESLint version:
```
v1.3.0
```

Source code:

```javascript
var number = 10;
var string = (number + 1) + 'px';
window.alert(string); // eslint-disable-line no-alert
```

ESLint output:
```
2:27  error  Unexpected string concatenation of literals  no-useless-concat
```

If I'm not mistaken then this concatenation is not useless because `number` is not a string.

### PR DESCRIPTION
**Pull request by Dashed**
Fixes #3575

Uses behaviour based on this description: https://github.com/eslint/eslint/issues/3575#issuecomment-136003933

Adapted from and is an alternative to PR https://github.com/eslint/eslint/pull/3579

----

@hzoo can you check if this patch is good enough?

### CODE REVIEW

File: lib/rules/no-useless-concat.js [hunk 1]
```diff
@@ -11,6 +11,34 @@
...
+    }
+}
+
+/**
+ * Checks whether or not a given binary expression has a string literal.
+ * @param {ASTNode} node - A node to check.
+ * @returns {boolean} `true` if the node has a string literal.
+ */
+function hasStringLiteral(node) {
+    if (node.type === "BinaryExpression") {
```
Comment (mysticatea) [inline lib/rules/no-useless-concat.js]: Here, if `node.operator` is not `"+"`, it's always non string.

File: lib/rules/no-useless-concat.js [hunk 2]
```diff
@@ -11,6 +11,34 @@
...
+        return hasStringLiteral(node.left) || hasStringLiteral(node.right);
+    }
+    if (node.type === "UnaryExpression") {
```
Comment (mysticatea) [inline lib/rules/no-useless-concat.js]: I guess `UnaryExpression` is always non string literal.

Comment (Dashed) [inline lib/rules/no-useless-concat.js]: Ah good point.

File: lib/rules/no-useless-concat.js (hunk 1 above)
Comment (Dashed) [inline lib/rules/no-useless-concat.js]: Ah good point.

**Comment by Dashed** [comment]
@mysticatea amended commit with your suggestions

**Comment by mysticatea** [comment]
I'm sorry, I'll close with #3602.

**Pull request closed without merging**